File: MobileMeteoTests/MMTMeteorogramFetchDelegateTests.m
//
//  MMTMeteorogramFetchDelegateTests.m
//  MobileMeteo
//
//  Created by Kamil Szostakowski on 08.06.2015.
//  Copyright (c) 2015 Kamil Szostakowski. All rights reserved.
//

#import <XCTest/XCTest.h>
#import <OCMock/OCMock.h>
#import "MMTMeteorogramFetchDelegate.h"

@interface MMTMeteorogramFetchDelegateTests : XCTestCase
@end

@implementation MMTMeteorogramFetchDelegateTests

#pragma mark - Test methods

- (void)testEmptyInitialization
{
    MMTMeteorogramFetchDelegate *delegate = [[MMTMeteorogramFetchDelegate alloc] initWithCompletion:nil];
    XCTAssertNil(delegate);
}

- (void)testInitialization
{
    MMTMeteorogramFetchDelegate *delegate = [[MMTMeteorogramFetchDelegate alloc] initWithCompletion:^(NSData *image, NSError *error){}];
    XCTAssertNotNil(delegate);
}

- (void)testFinishCallback
{
    XCTestExpectation *finishCallbackExpectation = [self expectationWithDescription:@"Finish callback expectation"];
    MMTMeteorogramFetchDelegate *delegate = [[MMTMeteorogramFetchDelegate alloc] initWithCompletion:^(NSData *image, NSError *error)
    {
        [finishCallbackExpectation fulfill];
    }];
    
    [delegate connectionDidFinishLoading:nil];
    
    [self waitForExpectationsWithTimeout:1 handler:nil];
}

- (void)testFinishCallbackResponse
{
    NSData *data = [NSData dataWithData:[NSMutableData dataWithLength:100]];
    XCTestExpectation *finishCallbackExpectation = [self expectationWithDescription:@"Finish callback expectation"];

    MMTMeteorogramFetchDelegate *delegate = [[MMTMeteorogramFetchDelegate alloc] initWithCompletion:^(NSData *image, NSError *error)
    {
        XCTAssertNil(error);
        XCTAssertNotNil(image);
        XCTAssertEqualObjects(image, data);
        [finishCallbackExpectation fulfill];
    }];
    
    [delegate connection:nil didReceiveData:data];
    [delegate connectionDidFinishLoading:nil];
    
    [self waitForExpectationsWithTimeout:1 handler:nil];
}

- (void)testFinishCallbackForMultipartResponse
{
    NSData *partOfResponse = [NSData dataWithData:[NSMutableData dataWithLength:100]];
    NSData *responseData = [NSData dataWithData:[NSMutableData dataWithLength:300]];
    XCTestExpectation *finishCallbackExpectation = [self expectationWithDescription:@"Finish callback expectation"];
    
    MMTMeteorogramFetchDelegate *delegate = [[MMTMeteorogramFetchDelegate alloc] initWithCompletion:^(NSData *image, NSError *error)
    {
        XCTAssertNil(error);
        XCTAssertNotNil(image);
        XCTAssertEqualObjects(image, responseData);
        [finishCallbackExpectation fulfill];
    }];
    
    [delegate connection:nil didReceiveData:partOfResponse];
    [delegate connection:nil didReceiveData:partOfResponse];
    [delegate connection:nil didReceiveData:partOfResponse];
    [delegate connectionDidFinishLoading:nil];
    
    [self waitForExpectationsWithTimeout:1 handler:nil];
}

- (void)testConnectionWillSendRequest
{
    NSURL *url = [NSURL URLWithString:@"http://www.meteo.pl/um/metco/mgram_pict.php?ntype=0u&row=367&col=227&lang=pl&fdate=2015060812"];
    NSURLRequest *request = [NSURLRequest requestWithURL:url];
    
    MMTMeteorogramFetchDelegate *delegate = [[MMTMeteorogramFetchDelegate alloc] initWithCompletion:^(NSData *image, NSError *error){}];
    NSURLRequest *req = [delegate connection:nil willSendRequest:request redirectResponse:nil];
    
    XCTAssertEqualObjects(url.absoluteString, req.URL.absoluteString);
}

- (void)testConnectionWillSendRedirectionRequest
{
    NSString *expectedUrl = @"http://www.meteo.pl/um/metco/mgram_pict.php?ntype=0u&row=367&col=227&lang=pl&fdate=2015060812";
    NSURL *url = [NSURL URLWithString:@"http://www.meteo.pl/um/php/mgram_search.php?NALL=53.585869&EALL=19.570815&lang=pl&fdate=2015060812"];
    
    NSHTTPURLResponse *response = [[NSHTTPURLResponse alloc] initWithURL:url statusCode:301 HTTPVersion:@"1.1" headerFields:@ {
        @"Location": @"./meteorogram_map_um.php?ntype=0u&row=367&col=227&lang=pl&fdate=2015060812"
    }];

    MMTMeteorogramFetchDelegate *delegate = [[MMTMeteorogramFetchDelegate alloc] initWithCompletion:^(NSData *image, NSError *error){}];
    NSURLRequest *request = [delegate connection:nil willSendRequest:nil redirectResponse:response];

    XCTAssertEqualObjects(expectedUrl, request.URL.absoluteString);
}

- (void)testConnectionFailedWithError
{
    XCTestExpectation *finishCallbackExpectation = [self expectationWithDescription:@"Finish callback expectation"];
    
    MMTMeteorogramFetchDelegate *delegate = [[MMTMeteorogramFetchDelegate alloc] initWithCompletion:^(NSData *image, NSError *error)
    {
        if (error)
            [finishCallbackExpectation fulfill];
    }];
    
    [delegate connection:nil didFailWithError:[NSError errorWithDomain:@"domain" code:1 userInfo:nil]];
    [self waitForExpectationsWithTimeout:1 handler:nil];
}

@end

File: MobileMeteoTests/MMTMeteorogramFetchDelegateTests.swift
//
//  MMTMeteorogramFetchDelegateTests.swift
//  MobileMeteo
//
//  Created by Kamil Szostakowski on 30.06.2015.
//  Copyright (c) 2015 Kamil Szostakowski. All rights reserved.
//

import XCTest
import Foundation
import CoreLocation
import MobileMeteo

class MMTMeteorogramFetchDelegateTests : XCTestCase
{
    // MARK: Properties
    
    var connection: NSURLConnection!
    var idleClosure: MMTFetchMeteorogramCompletion!
    
    // MARK: Setup methods
    
    override func setUp()
    {
        super.setUp();
        connection = NSURLConnection();
        idleClosure = {(var image: NSData?, var error: NSError?) in }
    }
    
    override func tearDown()
    {
        connection = nil;
        super.tearDown()
    }
    
    // MARK: Callback test methods
    
    func testFinishCallback()
    {
        let finishCallbackExpectation = expectationWithDescription("Finish callback expectation")
        let delegate = MMTMeteorogramRedirectionFetchDelegate(url: NSURL(), completion: {(var image: NSData?, var error: NSError?) in
            
            finishCallbackExpectation.fulfill()
        })
    
        delegate.connectionDidFinishLoading(NSURLConnection())
        waitForExpectationsWithTimeout(1, handler: nil)
    }
    
    func testFinishCallbackResponse()
    {
        let data = NSData(data: NSMutableData(length: 100)!)
        let finishCallbackExpectation = expectationWithDescription("Finish callback expectation")
    
        let delegate = MMTMeteorogramRedirectionFetchDelegate(url: NSURL(), completion: {(var image: NSData?, var error: NSError?) in
    
            XCTAssertNil(error);
            XCTAssertNotNil(image);
            XCTAssertEqual(image!, data);
        
            finishCallbackExpectation.fulfill();
        });
    
        delegate.connection(connection, didReceiveData: data)
        delegate.connectionDidFinishLoading(NSURLConnection())
        waitForExpectationsWithTimeout(1, handler: nil)
    }
    
    func testFinishCallbackForMultipartResponse()
    {
        let partOfResponse = NSData(data: NSMutableData(length: 100)!)
        let responseData = NSData(data: NSMutableData(length: 300)!)
        let finishCallbackExpectation = expectationWithDescription("Finish callback expectation")
    
        let delegate = MMTMeteorogramRedirectionFetchDelegate(url: NSURL()) {
            (var image: NSData?, var error: NSError?) in
    
            XCTAssertNil(error)
            XCTAssertNotNil(image)
            XCTAssertEqual(image!, responseData)
            finishCallbackExpectation.fulfill()
        };
    
        delegate.connection(connection, didReceiveData: partOfResponse)
        delegate.connection(connection, didReceiveData: partOfResponse)
        delegate.connection(connection, didReceiveData: partOfResponse)
        delegate.connectionDidFinishLoading(NSURLConnection())
        waitForExpectationsWithTimeout(1, handler: nil)
    }
    
    func testFinishCallbackForUnpreparedMeteorogram()
    {
        let response = NSData(data: NSMutableData(length: 71)!)
        let finishCallbackExpectation = expectationWithDescription("Finish callback expectation")
        
        let delegate = MMTMeteorogramRedirectionFetchDelegate(url: NSURL()) {
            (var image: NSData?, var error: NSError?) in
            
            XCTAssertNil(image)
            XCTAssertNotNil(error)
            XCTAssertEqual(error!.domain, MMTErrorDomain)
            XCTAssertEqual(error!.code, MMTError.MeteorogramNotFound.rawValue)
            
            finishCallbackExpectation.fulfill()
        }
        
        delegate.connection(connection, didReceiveData: response)
        delegate.connectionDidFinishLoading(NSURLConnection())
        waitForExpectationsWithTimeout(1, handler: nil)
    }
    
    // MARK: Delegate test methods    
    
    func testConnectionFailedWithError()
    {
        let finishCallbackExpectation = expectationWithDescription("Finish callback expectation")
        let delegate = MMTMeteorogramRedirectionFetchDelegate(url: NSURL()) {
            (var image: NSData?, var error: NSError?) in
            
            XCTAssertNotNil(error)
            XCTAssertEqual(error!.domain, MMTErrorDomain)
            XCTAssertEqual(error!.code, MMTError.MeteorogramFetchFailure.rawValue)
            
            finishCallbackExpectation.fulfill()
        };
    
        delegate.connection(connection, didFailWithError: NSError(domain: "domain", code: 1, userInfo: nil))
        waitForExpectationsWithTimeout(1, handler: nil)
    }
}